warning: control reaches end of non-void function [-Wreturn-type]

Building the latest code from the repo on Linux with cmake (which uses `-Wall -Wextra`) and g++ 11.4.0, I get these warnings:

```
.../CLK/Components/9918/Implementation/../Implementation/PersonalityTraits.hpp: In function ‘constexpr size_t TI::TMS::memory_size(TI::TMS::Personality)’:
.../CLK/Components/9918/Implementation/../Implementation/PersonalityTraits.hpp:41:1: warning: control reaches end of non-void function [-Wreturn-type]
   41 | }
      | ^
```

```
.../CLK/Machines/Acorn/Archimedes/../../../InstructionSets/ARM/Registers.hpp: In static member function ‘static constexpr uint32_t InstructionSet::ARM::Registers::pc_offset_during(InstructionSet::ARM::Registers::Exception)’:
.../CLK/Machines/Acorn/Archimedes/../../../InstructionSets/ARM/Registers.hpp:188:17: warning: control reaches end of non-void function [-Wreturn-type]
  188 |                 }
      |                 ^
```

I think this is because although you've handled all the enumeration cases, you haven't handled the hypothetical case of a some other integer being passed to these functions. While that situation probably doesn't arise, the warning could be silenced by adding a `default` case. In other functions I've seen you use:

https://github.com/TomHarte/CLK/blob/018f0e097fabe264695561e4e18cb46297cc083b/InstructionSets/M50740/Parser.hpp#L68

--- a/Components/9918/Implementation/PersonalityTraits.hpp
+++ b/Components/9918/Implementation/PersonalityTraits.hpp
@@ -38,6 +38,7 @@ constexpr size_t memory_size(Personality p) {
 		case TI::TMS::V9938:	return 128 * 1024;
 		case TI::TMS::V9958:	return 192 * 1024;
 	}
+	return 0;
 }
 
 constexpr size_t memory_mask(Personality p) {

--- a/InstructionSets/ARM/Executor.hpp
+++ b/InstructionSets/ARM/Executor.hpp
@@ -339,7 +339,9 @@ struct Executor {
 			if constexpr (flags.transfer_byte()) {
 				uint8_t target;
 				did_read = bus.template read<uint8_t>(address, target, registers_.mode(), trans);
-				value = target;
+				if(did_read) {
+					value = target;
+				}
 			} else {
 				did_read = bus.template read<uint32_t>(address, value, registers_.mode(), trans);
 

--- a/InstructionSets/ARM/Executor.hpp
+++ b/InstructionSets/ARM/Executor.hpp
@@ -337,7 +337,7 @@ struct Executor {
 			uint32_t value = 0;
 
 			if constexpr (flags.transfer_byte()) {
-				uint8_t target;
+				uint8_t target = 0;	// Value should never be used; this avoids a spurious GCC warning.
 				did_read = bus.template read<uint8_t>(address, target, registers_.mode(), trans);
 				if(did_read) {
 					value = target;

--- a/InstructionSets/ARM/Registers.hpp
+++ b/InstructionSets/ARM/Registers.hpp
@@ -185,6 +185,7 @@ struct Registers {
 				// Unspecified; a guess.
 				case Exception::Reset:					return 0;
 			}
+			return 4;
 		}
 
 		/// Updates the program counter, interupt flags and link register if applicable to begin @c exception.
@@ -270,10 +271,11 @@ struct Registers {
 				case Condition::GT:	return !le();
 				case Condition::LE:	return le();
 
-				default:
 				case Condition::AL:	return true;
 				case Condition::NV:	return false;
 			}
+
+			return false;
 		}
 
 		/// Sets current execution mode.
@@ -345,23 +347,19 @@ struct Registers {
 		/// this will the the user-mode register. Otherwise it'll be that for the current mode. These references
 		/// are guaranteed to remain valid only until the next mode change.
 		uint32_t &reg(bool force_user_mode, uint32_t offset) {
-			if(!force_user_mode) {
-				return active_[offset];
-			}
-
 			switch(mode_) {
 				default:
 				case Mode::User: return active_[offset];
 
 				case Mode::Supervisor:
 				case Mode::IRQ:
-					if(offset == 13 || offset == 14) {
+					if(force_user_mode && (offset == 13 || offset == 14)) {
 						return user_registers_[offset - 8];
 					}
 					return active_[offset];
 
 				case Mode::FIQ:
-					if(offset >= 8 && offset < 15) {
+					if(force_user_mode && (offset >= 8 && offset < 15)) {
 						return user_registers_[offset - 8];
 					}
 					return active_[offset];

--- a/Machines/Acorn/Archimedes/Archimedes.cpp
+++ b/Machines/Acorn/Archimedes/Archimedes.cpp
@@ -548,7 +548,7 @@ class ConcreteMachine:
 		}
 
 		uint32_t advance_pipeline(uint32_t pc) {
-			uint32_t instruction;
+			uint32_t instruction = 0;	// Value should never be used; this avoids a spurious GCC warning.
 			const bool did_read = executor_.bus.read(pc, instruction, executor_.registers().mode(), false);
 			return pipeline_.exchange(
 				did_read ? instruction : Pipeline::SWI,

--- a/Machines/Acorn/Archimedes/Archimedes.cpp
+++ b/Machines/Acorn/Archimedes/Archimedes.cpp
@@ -551,7 +551,7 @@ class ConcreteMachine:
 			uint32_t instruction;
 			const bool did_read = executor_.bus.read(pc, instruction, executor_.registers().mode(), false);
 			return pipeline_.exchange(
-				instruction,
+				did_read ? instruction : Pipeline::SWI,
 				did_read ? Pipeline::SWISubversion::None : Pipeline::SWISubversion::DataAbort);
 		}
 
@@ -562,6 +562,8 @@ class ConcreteMachine:
 				IRQ,
 				FIQ,
 			};
+
+			static constexpr uint32_t SWI = 0xef'000000;
 
 			uint32_t exchange(uint32_t next, SWISubversion subversion) {
 				const uint32_t result = upcoming_[active_].opcode;
@@ -586,7 +588,7 @@ class ConcreteMachine:
 			// In practice I got into a bit of a race condition between interrupt scheduling and
 			// flags changes, so have backed off for now.
 			void reschedule(SWISubversion subversion) {
-				upcoming_[active_].opcode = 0xef'000000;
+				upcoming_[active_].opcode = SWI;
 				upcoming_[active_].subversion = subversion;
 			}
 

--- a/Machines/Apple/AppleIIgs/MemoryMap.hpp
+++ b/Machines/Apple/AppleIIgs/MemoryMap.hpp
@@ -78,7 +78,7 @@ class MemoryMap {
 
 			const auto physical = physical_address(region, address);
 			assert(physical <= 0xff'ffff);
-			return shadow_pages_[(physical >> 10) & 127] & shadow_banks_[physical >> 17];
+			return shadow_pages_[(physical >> 10) & 127] && shadow_banks_[physical >> 17];
 		}
 		void write(const Region &region, uint32_t address, uint8_t value) {
 			if(!region.write) {

--- a/Machines/Apple/AppleIIgs/MemoryMap.hpp
+++ b/Machines/Apple/AppleIIgs/MemoryMap.hpp
@@ -77,7 +77,7 @@ class MemoryMap {
 			}
 
 			const auto physical = physical_address(region, address);
-			assert(physical >= 0 && physical <= 0xff'ffff);
+			assert(physical <= 0xff'ffff);
 			return shadow_pages_[(physical >> 10) & 127] & shadow_banks_[physical >> 17];
 		}
 		void write(const Region &region, uint32_t address, uint8_t value) {

--- a/Processors/68000/Implementation/68000Implementation.hpp
+++ b/Processors/68000/Implementation/68000Implementation.hpp
@@ -1869,6 +1869,7 @@ void Processor<BusHandler, dtack_is_implicit, permit_overrun, signal_will_perfor
 					SetupDataAccess(0, Operation::SelectWord);
 				MoveToStateSpecific(StoreOperand_l);
 
+				default:	// Convince GCC that nothing here is amiss.
 				case InstructionSet::M68k::DataSize::Word:
 					SetupDataAccess(0, Operation::SelectWord);
 				MoveToStateSpecific(StoreOperand_bw);

